diesel_cli/src/main.rs

```diff
@@ -103,69 +112,121 @@ fn run_migration_command(matches: &ArgMatches) {
     }
 }
 
-fn run_setup_command(matches: &ArgMatches) {
+fn run_database_command(matches: &ArgMatches) {
+    match matches.subcommand() {
+        ("setup", Some(args)) => setup_database(args).unwrap_or_else(handle_error),
+        ("reset", Some(args)) => reset_database(args).unwrap_or_else(handle_error),
+        _ => unreachable!("The cli parser should prevent reaching here"),
+    };
+}
+
+fn reset_database(args: &ArgMatches) -> Result<(), DatabaseError> {
+    let (database, postgres_url) = split_pg_connection_string(&database_url(args));
+    let connection = try!(PgConnection::establish(&postgres_url));
+    try!(drop_database(&connection, &database));
+    setup_database(args)
+}
+
+fn setup_database(args: &ArgMatches) -> Result<(), DatabaseError> {
     migrations::find_migrations_directory()
         .unwrap_or_else(|_|
                         create_migrations_directory()
                         .map_err(handle_error).unwrap()
                        );
 
-    if PgConnection::establish(&database_url(matches)).is_err() {
-        create_database(database_url(matches)).unwrap_or_else(handle_error);
+    let database_url = database_url(args);
+
+    if PgConnection::establish(&database_url).is_err() {
+        let (database, postgres_url) = split_pg_connection_string(&database_url);
+        let pg_connection = try!(PgConnection::establish(&postgres_url));
+        try!(create_database(&pg_connection, &database));
     }
 
-    let connection = connection(&database_url(matches));
+    let connection = connection(&database_url);
+    setup_schema_table(&connection)
+}
 
-    if !schema_table_exists(&connection).map_err(handle_error).unwrap() {
-        migrations::create_schema_migrations_table_if_needed(&connection)
-            .map_err(handle_error).unwrap();
-        migrations::run_pending_migrations(&connection)
-            .map_err(handle_error).unwrap();
-    }
+/// Creates the __diesel_schema_migrations table if it doesn't exist. If the
+/// table didn't exist, it also runs any pending migrations. Returns a
+/// `DatabaseError::ConnectionError` if it can't create the table, and exits
+/// with a migration error if it can't run migrations.
+fn setup_schema_table<Conn: Connection>(conn: &Conn) -> Result<(), DatabaseError> {
+    if !schema_table_exists(conn).map_err(handle_error).unwrap() {
+        try!(migrations::create_schema_migrations_table_if_needed(conn));
+        migrations::run_pending_migrations(conn).unwrap_or_else(handle_error);
+    };
+    Ok(())
+}
+
+/// Drops the database specified in the connection url. It returns a
+/// `DatabaseError::ConnectionError` if it can't connect to the postgres
```

Comment:
Yeah but how do we want to represent it in functions that deal with both a specific database's url, and the generic url?